### ISSUE TITLE
Certain file names appear differently (at the byte level) across systems

### ISSUE THREAD

**Issue by bnbalsamo** [uchicago-library/uchicagoldr-toolsuite#97]
While perhaps not part of the purview of the toolsuite to "fix" this issue drastically effects operations.

Certain filenames (eg: ones which use funky characters or ones that originate on antiquated/non-standard systems) present different when examined from different systems. This causes serious issues in all the file system serializations which rely on consistent file names, all the tools which rely on being able to encode file systems in utf-8 in logs and PREMIS records, and generally in terms of consistent object references.

We already consider the file name to be strictly metadata at the level of Archive structures (where it is only stored in the PREMIS record as a utf-8 encoded string). We need to move this mentality into the staging workflow at the earliest points in order to isolate the toolsuite from these issues of inconsistent file name encodings/interpretations.

This most likely will involve serious work in the staging workflow, as well as how objects are referenced in the structures themselves (MaterialSuites most likely need to be given an identifier).

Current thinking on this issue is that the file names should be treated as strictly a series of bytes, this series of bytes can then be encoded as a utf-8 string representation of a hexadecimal number. Interpretation of this encoding of the file name is left up to systems operating on the data in the future that have reason to need the original file name, and its interpretability is left suspect.

Serializations must then account for the use of a new opaque identifier (most likely a uuid4) to reference all files and the file name being stored in metadata. The end result of this will most likely be that the Stages end up looking significantly more similar to the identifier-heavy pairtree representations utilized in Archives.

**Comment by bnbalsamo** [comment]
See #93 - could this get lumped into a sanitization step in order to isolate it from the current operations of the stager? Would that be too "hacky" a fix?

While it might save us some work in reconfiguring staging code - I think the majority of the work will occur in the structures code base as well as the utilities that in some way currently rely (albeit usually tangentially) on the file names (the file system readers and writers, and some of the "quick rather than conceptually sound" mime-based workflows which are reading extensions.

With this in mind I think there is some discussion to be had about when and where it actually makes sense to do all of this stuff.

**Comment by verbalhanglider** [comment]
I agree. But, it's not just that material suites will need identifiers and other ways that that the objects are referenced. We will also need to create a "historical information" category in the admin section for each segment. 

This "historical information" will have have at least the following: 1. a snapshot of the directory structure of the stage as it was in origin; 2. a mapping file that takes the raw bytes string filename and links it to the uuid filename that is in staging. This will be necessary to facilitate the return of a deposit as it was when it arrived at the archives to a depositor.

To summarize the new workflow:
1. receive a file
2. copy file into ldr staging with a normalized name
3. add a mapping of original file name as binary string to the new normalized name
4. process the file in staging life cycle
5. rename file again to longTermStorage naming convention when it is archived
6. retain mapping in "historical information" section of admin environment for the accession in longTermStorage

**Comment by bnbalsamo** [comment]
I think it might worth also exploring this as an opportunity to move the incoming file into the back seat immediately on ingestion, which previously was undesirable because of the thought that we could leverage the original file name/directory structure to keep things in order.

I think this assumption came about as a natural progression from our original rsync/fits workflows, but has now proven to be a bigger liability than help as we've progressed.

If we create a (minimal) PREMIS record on ingest, I think we can implicitly store the state of the directory structure and our name <--> uuid link without introducing a new identifier, and also leverage PREMIS for some of the business we are utilizing the stand alone manifest for as well.

Thus, on ingest a PREMIS record would be created containing _only_ the required elements + our hex encoded original name string

```
premis
    object
        objectIdentifier
            objectIdentifierValue --> uuid4()
            objectIdentifierType --> "uuid4"
        objectCharacteristics
            format
                formatDesignation --> detect_mime(copied_file)
        originalName --> hex_encode_filepath_bytes(copied_file.rel_path)
```

Then, our original directory structure (which we can't serialize safely in a cross platform manner for the same reason we can't safely serialize file names) becomes the aggregate of the originalName fields, interpreted from hex encoded bytes in whatever encoding the interested party has determined the file names to be written in.

The file bytes get deposited in a pairtree ala the archive structure, at the end of a pairtree branch of their PREMIS identifier.

Segments can then be serialized as _entirely_ administrative information, with a new administrative component that is a list of PREMIS identifiers that corresponds to the contained files.

**Comment by bnbalsamo** [comment]
Just going to type out a hypothesized file level workflow for the stager so that I have one around for the meeting this afternoon
1. Read the file name from the origin media - convert it to hex, store it in RAM
2. Copy the file from the origin into a temporary location
3. Try to compute the [sha256? | md5? | adler32? | this trade off is important] for the file on the origin media
4. Compute the same fixity metric as above on the temporary location
5. Document fixity result (allow defining behavior on mismatches here?)
6. Generate a stub PREMIS record (including remaining fixity metrics) from file now safely on our hardware, including hex representation of file name bytes
7. Throw references to both the file content and the PREMIS record into the structure in RAM
8. Wait for the writer to be called on the structure and then serialize the new data into the staging environment itself along with its freshly minted PREMIS record

**Comment by bnbalsamo** [comment]
Work has proceeded and PREMIS is now considered to be the canonical requirement in a MaterialSuite (instead of content) which has allowed us to the move file names entirely into metadata, and decouple serializations from them entirely while still storing them.

In regards to how the file names are stored in records I've opted to attempt to leave them more human readable (if making decoding slightly more complicated) utilizing the built-in stdlib methods [fsencode()](https://docs.python.org/3/library/os.html#os.fsencode) and [fsdecode()](https://docs.python.org/3/library/os.html#os.fsdecode). This makes the originalName field _significantly_ more human readable, but does rely on being able to "guess" the file system encoding used on the machine which generated the PREMIS in the future. Because _nearly_ all modern OSs use some form of UTF (8 or 16) as the default file system encoding this probably won't end up a tremendous issue in the future. (knock on wood)

With this change I'm content to call this issue closed once the new staging serializations/PREMIS based paradigms have been merged into dev.

**Comment by bnbalsamo** [comment]
This issue _should_ be remedied, as of the merge of #103 into dev
